### PR TITLE
FIX bug that makes clean data not to work on app exit when VPN enabled

### PR DESCRIPTION
**Pull request by aitorvs**
Task/Issue URL: https://app.asana.com/0/414730916066338/1201439663288660/f

### Description
With AppTP we introduced a bug that manifests itself when the AppTP is ENABLED

_Repro steps_
* install app and enable appTP
* open couple tabs
* go to Settings → Automatic clear
  * set Tabs and Data
* Close app and re-open
* Expected
  * data is clear
* Actual
  * data is not clear

The problem is in the logic in `AutomaticDataClearer`. The class keeps some state, in particular the `isFreshAppLaunch` field. That it is assumed to be set back to `true` when the app is closed and re-open.
This is becasue the `AutomaticDataClearer` assumes/expects that Android will kill the app process when the user swipes the app closed, however
* this is true when the VPN service is OFF
* but it is NOT true when the VPN service is ON
  * in that case, swiping the app closed will NOT kill the app process
  * `AutomaticDataClearer` instance is not cleaned and so the `isFreshAppLaunch` will remain set to `false` not clearing the data

_Solution_
The only reliable way to detect whether the app is closed with and without VPN enabled is to track the app activities states. When the app does not have any activity in CREATED state, we consider the app is closed.
* a new class `BrowserApplicationStateInfo` that performs that tracking
* and calls `BrowserLifecycleObserver` observer callbacks
  * `onOpen`, `onForeground`, `onBackground`, `onClosed` and `onExit`
  * see javadocs in `BrowserLifecycleObserver` class to understahdn when they are called

_Why the new gradle module browser-api?_
* I considered that `BrowserLifecycleObserver` would also be handy in other scenarios as oppose to rely on `LifecycleObserver`
* as `BrowserLifecycleObserver` may get used from other gradle modules I already advanced that and created the `browser-api` module
  * `browser-api` module will be home for any public API from the browser app
  * `browser-api` was already planned to be added under [this project](https://app.asana.com/0/72649045549333/1199214002677571/f). This PR just advances that a bit

### Steps to test this PR

_Clear tabs and data_
1. install from this branch
1. go to settings -> automatic clear and configure to clear "tabs and data" on "app exit"
1. open couple tabs
1. in one of the tabs, search for eg. "hola" and click in one of the links
1. open another tab and visit https://httpbin.org/basic-auth/u/p
  * login with user: u and password: p
1. home the app and reopen
1. VERIFY clear didn't happen and tabs are still there
1. swipe close the app and reopen
1. VERIFY tabs are not there
1. search for "hola" and verify that the previously opened link is not marked as visited
1. visit https://httpbin.org/basic-auth/u/p
1. VERIFY you will need to input the login credentials again

_Clear tabs_
1. install from this branch
1. go to settings -> automatic clear and configure to clear "tabs" on "app exit"
1. open couple tabs
1. in one of the tabs, search for eg. "hola" and click in one of the links
1. open another tab and visit https://httpbin.org/basic-auth/u/p
  * login with user: u and password: p
1. home the app and reopen
1. VERIFY clear didn't happen and tabs are still there
1. swipe close the app and reopen
1. VERIFY tabs are not there
1. search for "hola" and verify that the previously opened link is not marked as visited
1. visit https://httpbin.org/basic-auth/u/p
1. VERIFY you will NOT need to input the login credentials again

### CODE REVIEW

**Comment by aitorvs** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #1593**  (this PR)

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh) and will continue to be automatically updated while this PR remains open.

**Comment by aitorvs** [comment]
Some unit tests may fail but they pass when I do it manually. Looking into that but opened the PR for early review while I investigate the unit test

File: app/src/main/java/com/duckduckgo/app/browser/state/BrowserApplicationStateInfo.kt [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+@ContributesMultibinding(AppObjectGraph::class)
+@Singleton
+class BrowserApplicationStateInfo @Inject constructor(
+    private val observers: Set<@JvmSuppressWildcards BrowserLifecycleObserver>
+) : ActivityLifecycleCallbacks {
+    private var created = 0
+    private var started = 0
+    private var resumed = 0
+
+    private val freshLaunchQueue: Queue<Boolean> = LinkedList()
```
Comment (CDRussell) [inline app/src/main/java/com/duckduckgo/app/browser/state/BrowserApplicationStateInfo.kt:38]: why is this a queue of `Boolean` vs a `Boolean`?

Comment (aitorvs) [reply app/src/main/java/com/duckduckgo/app/browser/state/BrowserApplicationStateInfo.kt:38]: no particular reason, in the beginning I though I would need to queue more events and then forgot to make it `Boolean?`. Will change

**Comment by aitorvs** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #1593**  (this PR)

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh) and will continue to be automatically updated while this PR remains open.

File: app/src/main/java/com/duckduckgo/app/fire/AutomaticDataClearer.kt [hunk 2]
```diff
@@ -122,6 +133,11 @@ class AutomaticDataClearer(
         }
     }
 
+    override fun onExit() {
+        // the app does not have any activity in CREATED state we kill the process
+        clearDataAction.killProcess()
```
Comment (CDRussell) [inline app/src/main/java/com/duckduckgo/app/fire/AutomaticDataClearer.kt:138]: This doesn't seem right. This indiscriminately kills the process regardless of the automatic data clearing settings.

One problem this creates if is you resize the window, then drag it to be full screen again. It triggers the process restart even though (to the user) the app was always in the foreground.

**Comment by aitorvs** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #1593**  (this PR)

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh) and will continue to be automatically updated while this PR remains open.

**Comment by aitorvs** [comment]
I eventually had to open the BrowserActivity because we're doing the save cast to it in the BrowserApplicationStateInfo. So we really need to pass an instance of BrowserActivity

**Comment by aitorvs** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #1593**  (this PR)

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh) and will continue to be automatically updated while this PR remains open.

**Review by CDRussell — APPROVED** [review]
This is a good workaround for a nasty problem 👍

**Comment by aitorvs** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #1593**  (this PR)

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh) and will continue to be automatically updated while this PR remains open.

**Comment by aitorvs** [comment]
@cdrussell take a look at the last commit. I had to disabled the Espresso tests because as they create the BrowserActivity, when destroyed the process is killed failing the tests. I tried to replace the dependencies like we do with eg. StubAppConfigurationDownloadModule but it didn't work. Still not sure why it didn't but I kind of gave up for now. I plan to take another look at it but proposing to merge as is. Will create an asana task

**Comment by aitorvs** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #1593**  (this PR)

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh) and will continue to be automatically updated while this PR remains open.

**Comment by aitorvs** [comment]
@cdrussell you suggestion worked. No need to ignore the esperesso tests anymore

**Pull request merged**